contracts/IController.sol

```diff
@@ -1,17 +1,18 @@
-pragma solidity ^0.5.11;
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.8;
 
 import "./zeppelin/Pausable.sol";
 
-contract IController is Pausable {
+abstract contract IController is Pausable {
     event SetContractInfo(bytes32 id, address contractAddress, bytes20 gitCommitHash);
 
     function setContractInfo(
         bytes32 _id,
         address _contractAddress,
         bytes20 _gitCommitHash
-    ) external;
+    ) external virtual;
 
-    function updateController(bytes32 _id, address _controller) external;
+    function updateController(bytes32 _id, address _controller) external virtual;
 
-    function getContract(bytes32 _id) public view returns (address);
+    function getContract(bytes32 _id) external view virtual returns (address);
```

Comment:
Could the visibility of `getContract()` stay as `public` here since the implementation in `Controller` is public?

Reply:
my bad, I first tried converting the contract to an interface when I added the external keyword, then did not change it back to public when I made it abstract again. fixed in ebbb1ef